File: backend/src/routes/systems.ts
import { Router, Response } from 'express';
import { authenticate, AuthRequest, requireRole } from '../middleware/auth';
import { AiSystem, GovernanceRule, Decision, BiasFlag } from '../models';
import { calculateHealthScore } from '../services/healthScore';

const router = Router();

// GET all systems
router.get('/', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const page = parseInt(req.query.page as string) || 1;
    const limit = parseInt(req.query.limit as string) || 10;
    
    // Using lean() for speed, map _id to id for frontend compatibility
    const systemsRaw = await AiSystem
      .find({ isActive: true })
      .sort({ createdAt: -1 })
      .skip((page - 1) * limit)
      .limit(limit)
      .lean();

    const systems = systemsRaw.map(s => ({
      ...s,
      id: s._id.toString()
    }));

    const total = await AiSystem.countDocuments({ isActive: true });

    return res.json({
      data: systems,
      meta: {
        total,
        page,
        limit,
        totalPages: Math.ceil(total / limit)
      }
    });
  } catch (error) {
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// POST new system (ADMIN)
router.post('/', authenticate, requireRole(['ADMIN']), async (req: AuthRequest, res: Response) => {
  try {
    const { 
      name, domain, description, 
      accuracy, precision, recall, fairnessScore, trainingDatasetSize 
    } = req.body;
    
    const system = await AiSystem.create({ 
      name, domain, description,
      accuracy, precision, recall, fairnessScore, trainingDatasetSize
    });
    return res.status(201).json(system);
  } catch (error) {
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// GET single system
router.get('/:id', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const system = await AiSystem.findById(req.params.id);
    if (!system) return res.status(404).json({ error: 'System not found' });
    
    // Note: the original prisma call used `include: { rules: true }`. We can't do that simply with refs 
    // unless defined, but the frontend likely hits /:id/rules for rules based on the routes below.
    // We attach rules manually if the frontend actually expects it nested here:
    const rules = await GovernanceRule.find({ aiSystemId: req.params.id, isActive: true });
    
    // Mongoose toJSON virtuals are activated inside res.json(system) but since we must insert rules:
    return res.json({
      ...system.toJSON(),
      rules
    });
  } catch (error) {
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// PATCH system (ADMIN)
router.patch('/:id', authenticate, requireRole(['ADMIN']), async (req: AuthRequest, res: Response) => {
  try {
    const { 
      name, domain, description, isActive,
      accuracy, precision, recall, fairnessScore, trainingDatasetSize
    } = req.body;
    
    const system = await AiSystem.findByIdAndUpdate(
      req.params.id,
      { $set: { 
        name, domain, description, isActive,
        accuracy, precision, recall, fairnessScore, trainingDatasetSize
      } },
      { new: true } // Return updated document
    );
    if (!system) return res.status(404).json({ error: 'System not found' });
    
    return res.json(system);
  } catch (error) {
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// GET system rules
router.get('/:id/rules', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const rulesRaw = await GovernanceRule.find({
      aiSystemId: req.params.id,
      isActive: true
    }).lean();

    const rules = rulesRaw.map(r => ({
      ...r,
      id: r._id.toString()
    }));

    return res.json(rules);
  } catch (error) {
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// POST system rules (ADMIN)
router.post('/:id/rules', authenticate, requireRole(['ADMIN']), async (req: AuthRequest, res: Response) => {
  try {
    const { name, description, category } = req.body;
    const rule = await GovernanceRule.create({
      aiSystemId: req.params.id as any,
      name,
      description,
      category
    });
    return res.status(201).json(rule);
  } catch (error) {
    return res.status(500).json({ error: 'Internal server error' });
  }
});

// GET system health score
router.get('/:id/health', authenticate, async (req: AuthRequest, res: Response) => {
  try {
    const aiSystemId = req.params.id as any;
    // We limit to 50 instead of fetching everything indefinitely
    const decisions = await Decision.find({ aiSystemId })
      .sort({ createdAt: -1 })
      .limit(50)
      .lean();

    if (decisions.length === 0) {
       return res.json({ score: 0, grade: 'Needs Data', trend: 'stable', metrics: null });
    }

    // We must manually aggregate bias flags or fetch them if they aren't embedded
    const decisionIds = decisions.map(d => d._id);
    const biasFlags = await BiasFlag.find({ decisionId: { $in: decisionIds } }).lean();

    let totalCompliance = 0;
    let totalTransparency = 0;
    let blockedCount = 0;
    let totalBiasFlags = biasFlags.length;
    let correctedBiasFlags = biasFlags.filter(b => b.corrected).length;

    decisions.forEach((d: any) => {
       totalCompliance += d.ethicalComplianceRate;
       totalTransparency += d.transparencyIndex;
       if (d.status === 'BLOCKED') blockedCount++;
    });

    const avgCompliance = totalCompliance / decisions.length;
    const avgTransparency = totalTransparency / decisions.length;
    const avgCog = decisions.reduce((acc: number, d: any) => acc + (d.cognitiveConsistency || 0), 0) / decisions.length;
    const notBlockedRatio = 1 - (blockedCount / decisions.length);
    const correctedRatio = totalBiasFlags === 0 ? 1 : (correctedBiasFlags / totalBiasFlags);

    const score = calculateHealthScore({
      ethicalComplianceRate: avgCompliance,
      transparencyIndex: avgTransparency,
      cognitiveConsistency: avgCog,
      selfRepairEfficiency: correctedRatio
    });
    
    let grade = 'Needs Review';
    if (score >= 90) grade = 'Excellent';
    else if (score >= 75) grade = 'Good';
    else if (score < 50) grade = 'Critical';

    const trend = score >= 85 ? 'improving' : (score < 60 ? 'declining' : 'stable');

    return res.json({
       score: Math.round(score),
       grade,
       trend,
       metrics: {
         avgCompliance: Math.round(avgCompliance * 100),
         avgTransparency: Math.round(avgTransparency * 100),
         notBlockedRatio: Math.round(notBlockedRatio * 100),
         correctedRatio: Math.round(correctedRatio * 100)
       }
    });
  } catch(error) {
     return res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: backend/routes/systems.py
"""Systems routes – replaces routes/systems.ts.

GET    /api/systems
POST   /api/systems
GET    /api/systems/:id
PATCH  /api/systems/:id
GET    /api/systems/:id/rules
POST   /api/systems/:id/rules
GET    /api/systems/:id/health
"""

from flask import Blueprint, request, jsonify, g
from bson import ObjectId
from middleware.auth import login_required, require_role
from models import serialize_doc, serialize_docs
from models import ai_system as SystemModel
from models import governance_rule as GovModel
from models import decision as DecisionModel
from models import bias_flag as BiasFlagModel
from services.health_score import calculate_health_score

bp = Blueprint("systems", __name__, url_prefix="/api/systems")


@bp.route("/", methods=["GET"])
@login_required
def list_systems():
    try:
        page = int(request.args.get("page", 1))
        limit = int(request.args.get("limit", 10))
        systems, total = SystemModel.find_active(page, limit)
        return jsonify({
            "data": serialize_docs(systems),
            "meta": {
                "total": total,
                "page": page,
                "limit": limit,
                "totalPages": -(-total // limit),
            },
        })
    except Exception as e:
        print(f"List systems error: {e}")
        return jsonify({"error": "Internal server error"}), 500


@bp.route("/", methods=["POST"])
@login_required
@require_role(["ADMIN"])
def create_system():
    try:
        body = request.get_json(force=True)
        system = SystemModel.create(body)
        return jsonify(serialize_doc(system)), 201
    except Exception:
        return jsonify({"error": "Internal server error"}), 500


@bp.route("/<system_id>", methods=["GET"])
@login_required
def get_system(system_id):
    try:
        system = SystemModel.find_by_id(system_id)
        if not system:
            return jsonify({"error": "System not found"}), 404
        rules = GovModel.find_active_by_system(system_id)
        doc = serialize_doc(system)
        doc["rules"] = serialize_docs(rules)
        return jsonify(doc)
    except Exception:
        return jsonify({"error": "Internal server error"}), 500


@bp.route("/<system_id>", methods=["PATCH"])
@login_required
@require_role(["ADMIN"])
def update_system(system_id):
    try:
        body = request.get_json(force=True)
        updates = {}
        for key in ["name", "domain", "description", "isActive",
                     "accuracy", "precision", "recall", "fairnessScore",
                     "trainingDatasetSize"]:
            if key in body:
                updates[key] = body[key]

        system = SystemModel.update_by_id(system_id, updates)
        if not system:
            return jsonify({"error": "System not found"}), 404
        return jsonify(serialize_doc(system))
    except Exception:
        return jsonify({"error": "Internal server error"}), 500


@bp.route("/<system_id>/rules", methods=["GET"])
@login_required
def get_rules(system_id):
    try:
        rules = GovModel.find_active_by_system(system_id)
        return jsonify(serialize_docs(rules))
    except Exception:
        return jsonify({"error": "Internal server error"}), 500


@bp.route("/<system_id>/rules", methods=["POST"])
@login_required
@require_role(["ADMIN"])
def create_rule(system_id):
    try:
        body = request.get_json(force=True)
        rule = GovModel.create({
            "aiSystemId": system_id,
            "name": body["name"],
            "description": body.get("description", ""),
            "category": body["category"],
        })
        return jsonify(serialize_doc(rule)), 201
    except Exception:
        return jsonify({"error": "Internal server error"}), 500


@bp.route("/<system_id>/health", methods=["GET"])
@login_required
def get_health(system_id):
    try:
        decisions = DecisionModel.find_by_system(system_id, limit=50)
        if not decisions:
            return jsonify({"score": 0, "grade": "Needs Data", "trend": "stable", "metrics": None})

        decision_ids = [d["_id"] for d in decisions]
        bias_flags = BiasFlagModel.find_by_decisions(decision_ids)

        total_compliance = sum(d["ethicalComplianceRate"] for d in decisions)
        total_transparency = sum(d["transparencyIndex"] for d in decisions)
        blocked_count = sum(1 for d in decisions if d["status"] == "BLOCKED")
        total_bias = len(bias_flags)
        corrected_bias = sum(1 for b in bias_flags if b.get("corrected"))

        n = len(decisions)
        avg_compliance = total_compliance / n
        avg_transparency = total_transparency / n
        avg_cog = sum(d.get("cognitiveConsistency", 0) for d in decisions) / n
        not_blocked_ratio = 1 - (blocked_count / n)
        corrected_ratio = 1 if total_bias == 0 else corrected_bias / total_bias

        score = calculate_health_score({
            "ethicalComplianceRate": avg_compliance,
            "transparencyIndex": avg_transparency,
            "cognitiveConsistency": avg_cog,
            "selfRepairEfficiency": corrected_ratio,
        })

        if score >= 90:
            grade = "Excellent"
        elif score >= 75:
            grade = "Good"
        elif score < 50:
            grade = "Critical"
        else:
            grade = "Needs Review"

        trend = "improving" if score >= 85 else ("declining" if score < 60 else "stable")

        return jsonify({
            "score": round(score),
            "grade": grade,
            "trend": trend,
            "metrics": {
                "avgCompliance": round(avg_compliance * 100),
                "avgTransparency": round(avg_transparency * 100),
                "notBlockedRatio": round(not_blocked_ratio * 100),
                "correctedRatio": round(corrected_ratio * 100),
            },
        })
    except Exception as e:
        print(f"Health error: {e}")
        return jsonify({"error": "Internal server error"}), 500
